File: BlotoutAnalytics/BlotoutAnalyticsAll/Controllers/BOASDKManifestController.m
//
//  BOASDKManifestController.m
//  BlotoutAnalytics
//
//  Copyright © 2019 Blotout. All rights reserved.
//

/**
 * The BOASDKManifestController is Controller to fetch SDK config from server
 */

#import "BOASDKManifestController.h"
#import <BlotoutFoundation/BOFNetworkPromise.h>
#import <BlotoutFoundation/BOFNetworkPromiseExecutor.h>
#import <BlotoutFoundation/BOFLogs.h>
#import "BOAUtilities.h"
#import <BlotoutFoundation/BOFFileSystemManager.h>
#import <BlotoutFoundation/BOFUserDefaults.h>
#import "BlotoutAnalytics_Internal.h"
#import "BOANetworkConstants.h"
#import "BOManifestAPI.h"
#import "NSError+BOAdditions.h"
#import "BOEventsOperationExecutor.h"

static id sBOAsdkManifestSharedInstance = nil;

@interface BOASDKManifestController ()

@end

@implementation BOASDKManifestController

-(instancetype)init {
  self = [super init];
  return self;
}

/**
 * method to get the singleton instance of the BOASDKManifestController object,
 * @return BOASDKManifestController instance
 */
+ (nullable instancetype)sharedInstance {
  static dispatch_once_t boaSDKManifestOnceToken = 0;
  dispatch_once(&boaSDKManifestOnceToken, ^{
    sBOAsdkManifestSharedInstance = [[[self class] alloc] init];
  });
  return sBOAsdkManifestSharedInstance;
}

/**
 * sync manifest and reload menifest data
 */
-(void)serverSyncManifestAndAppVerification:(void (^_Nullable) (BOOL isSuccess, NSError* error))callback {
  @try {
    [self fetchAndPrepareSDKModelWith:^(BOOL isSuccess, NSError *error) {
      if (isSuccess) {
        [self reloadManifestData];
      }
      
      if (!self.sdkManifestModel) {
        NSError *manifestReadError = nil;
        BOASDKManifest *sdkManifestM = [BOASDKManifest fromJSON:[self latestSDKManifestJSONString] encoding:NSUTF8StringEncoding error:&manifestReadError];
        self.sdkManifestModel = sdkManifestM;
      }
      callback(isSuccess, error);
    }];
  } @catch (NSException *exception) {
    BOFLogDebug(@"%@:%@", BOA_DEBUG, exception);
    callback(NO, [NSError boErrorForDict:exception.userInfo]);
  }
}

/**
 * method to get sdk manifest path
 * @return sdkManifestFilePath as string
 */
-(NSString*)latestSDKManifestPath {
  @try {
    NSString *fileName = @"sdkManifest";
    NSString *sdkManifestDir = [BOFFileSystemManager getSDKManifestDirectoryPath];
    NSString *sdkManifestFilePath = [NSString stringWithFormat:@"%@/%@.txt",sdkManifestDir, fileName];
    return sdkManifestFilePath;
  } @catch (NSException *exception) {
    BOFLogDebug(@"%@:%@", BOA_DEBUG, exception);
  }
  return nil;
}

/**
 * method to get manifest json string
 * @return sdkManifestStr as string
 */
-(NSString*)latestSDKManifestJSONString {
  @try {
    NSString *sdkManifestFilePath = [self latestSDKManifestPath];
    NSError *fileReadError;
    NSString *sdkManifestStr = [BOFFileSystemManager contentOfFileAtPath:sdkManifestFilePath withEncoding:NSUTF8StringEncoding andError:&fileReadError];
    return sdkManifestStr;
  } @catch (NSException *exception) {
    BOFLogDebug(@"%@:%@", BOA_DEBUG, exception);
  }
  return nil;
}

/**
 * method to get manifest path after writing
 * @param sdkManifest as String
 * @return sdkManifestFilePath as string
 */
-(NSString*)sdkManifestPathAfterWriting:(NSString*)sdkManifest {
  @try {
    if (!sdkManifest || [sdkManifest isEqualToString:@""] || [sdkManifest isEqualToString:@"{}"] || [sdkManifest isEqualToString:@"{ }"]) {
      return nil;
    }
    
    NSString *sdkManifestFilePath = [self latestSDKManifestPath];
    if ([BOFFileSystemManager isFileExistAtPath:sdkManifestFilePath]) {
      NSError *removeError = nil;
      [BOFFileSystemManager removeFileFromLocationPath:sdkManifestFilePath removalError:&removeError];
    }
    
    NSError *error;
    //else file write operation and prapare new object
    [BOFFileSystemManager pathAfterWritingString:sdkManifest toFilePath:sdkManifestFilePath writingError:&error];
    
    return sdkManifestFilePath;
  } @catch (NSException *exception) {
    BOFLogDebug(@"%@:%@", BOA_DEBUG, exception);
  }
  return nil;
}

/**
 * method to fetch and prepare model for sdk manifest config
 */
-(void)fetchAndPrepareSDKModelWith:(void (^_Nullable) (BOOL isSuccess, NSError* error))callback {
  @try {
    BOManifestAPI *api = [[BOManifestAPI alloc] init];
    
    [api getManifestDataModel:^(id  _Nonnull responseObject, id  _Nonnull data) {
      if (!responseObject) {
        self.isSyncedNow = NO;
        callback(NO, nil);
        return;
      }
      
      BOASDKManifest *sdkManifestM = (BOASDKManifest*)responseObject;
      self.sdkManifestModel = sdkManifestM;
      NSString *manifestJSONStr = [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
      [self sdkManifestPathAfterWriting:manifestJSONStr];
      self.isSyncedNow = YES;
      callback(YES, nil);
    } failure:^(NSError * _Nonnull error) {
      self.isSyncedNow = NO;
      callback(NO, error);
    }];
  } @catch (NSException *exception) {
    BOFLogDebug(@"%@:%@", BOA_DEBUG, exception);
    callback(NO, [NSError boErrorForDict:exception.userInfo]);
  }
}

/**
 * method to get manifest variable
 * @param ID as String
 * @return oneVar as BOASDKVariable
 */
-(BOASDKVariable*)getManifestVariable:(BOASDKManifest*)manifest forID:(int)ID {
  @try {
    BOASDKVariable *oneVar = nil;
    for (BOASDKVariable *oneVariableDict in manifest.variables) {
      if (oneVariableDict != nil && [oneVariableDict.variableID intValue] == ID) {
        oneVar = oneVariableDict;
        break;
      }
    }
    return oneVar;
  } @catch (NSException *exception) {
    BOFLogDebug(@"%@:%@", BOA_DEBUG, exception);
  }
  return nil;
}

/**
 * method to manifest data
 */
-(void)reloadManifestData {
  @try {
    NSString *manifestStr = [self latestSDKManifestJSONString];
    if (!manifestStr) {
      return;
    }
    
    if (self.sdkManifestModel == nil && manifestStr != nil && ![manifestStr isEqualToString: @""]) {
      NSError *manifestReadError = nil;
      BOASDKManifest *sdkManifestM = [BOASDKManifest fromJSON: manifestStr encoding:NSUTF8StringEncoding error:&manifestReadError];
      self.sdkManifestModel = sdkManifestM;
    }
    
    if (self.sdkManifestModel == nil) {
      return;
    }
    
    BOASDKVariable *systemEvents = [self getManifestVariable:self.sdkManifestModel forID:MANIFEST_SYSTEM_EVENTS];
    if (systemEvents != nil) {
      self.enabledSystemEvents = [systemEvents.value componentsSeparatedByString:@","];
    }
    
    BOASDKVariable *piiKey = [self getManifestVariable:self.sdkManifestModel forID:MANIFEST_PII_PUBLIC_KEY];
    if (piiKey != nil) {
      self.piiPublicKey = piiKey.value;
    }
    
    BOASDKVariable *phiKey = [self getManifestVariable:self.sdkManifestModel forID:MANIFEST_PHI_PUBLIC_KEY];
    if (phiKey != nil) {
      self.phiPublickey = phiKey.value;
    }
  } @catch (NSException *exception) {
    BOFLogDebug(@"%@:%@", BOA_DEBUG, exception);
  }
}

/**
 * method to check weather manifest is available or not
 * @return status as BOOL
 */
-(BOOL)isManifestAvailable {
  @try {
    NSString *manifestStr = [self latestSDKManifestJSONString];
    if (manifestStr && ![manifestStr isEqualToString:@""]) {
        return YES;
    }
    return NO;
  } @catch (NSException *exception) {
    BOFLogDebug(@"%@:%@", BOA_DEBUG, exception);
  }
  return NO;
}

-(BOOL)isSystemEventEnabled:(int)eventCode {
  if (self.enabledSystemEvents == nil) {
    return false;
  }
  
  return [self.enabledSystemEvents containsObject: [@(eventCode) stringValue]];
}

@end

File: BlotoutAnalytics/Controllers/BOASDKManifestController.swift
//
//  BOASDKManifestController.swift
//  BlotoutAnalyticsSDK
//
//  Created by Poonam Tiwari on 23/03/22.
//

import Foundation

private var sBOAsdkManifestSharedInstance: Any? = nil

class BOASDKManifestController:NSObject {
    
    var sdkManifestModel: BOASDKManifest?
    
    //newly added
    var manifestModel:ManifestModel?
    
    var isSyncedNow = false
    var piiPublicKey: String?
    var phiPublickey: String?
    var enabledSystemEvents: [AnyHashable]?
    
    static let sharedInstance = BOASDKManifestController()
    override init() {
        super.init()
    }
    
//    class func sharedInstance() -> Self {
//        // TODO: [Swiftify] ensure that the code below is executed only once (`dispatch_once()` is deprecated)
//        { [self] in
//            sBOAsdkManifestSharedInstance = self.init()
//        }
//        return sBOAsdkManifestSharedInstance as! Self
//    }
    
    func serverSyncManifestAndAppVerification(_ callback:@escaping ((_ isSuccess: Bool, _ error: Error?) -> Void)) {
        
        fetchAndPrepareSDKModel() { isSuccess, error in
            
            if isSuccess {
                self.reloadManifestData()
            }
            
            //TODO: have removed model null & refill scenario, add again if needed

            callback(isSuccess,error)
        }
    }
    
    
    func latestSDKManifestPath() -> String {
        
        let fileName = "sdkManifest"
        let sdkManifestDir = BOFFileSystemManager.getSDKManifestDirectoryPath()
        let sdkManifestFilePath = "\(sdkManifestDir ?? "")/\(fileName).txt"
        return sdkManifestFilePath
    }
    
    func latestSDKManifestJSONString() -> String? {
       do{
            let sdkManifestFilePath = latestSDKManifestPath()
            var fileReadError: Error?
           let sdkManifestStr = try BOFFileSystemManager.contentOfFile(atPath: sdkManifestFilePath, with: String.Encoding.utf8)//, andError: &fileReadError)
            return sdkManifestStr
        } catch {
            BOFLogDebug(frmt: "%@:%@", args: BOF_DEBUG, error.localizedDescription)
        }
        return nil
    }
    
    func sdkManifestPath(afterWriting sdkManifest: String) -> String? {
       do{
            if sdkManifest == nil || (sdkManifest == "") || (sdkManifest == "{}") || (sdkManifest == "{ }") {
                return nil
            }
            
            let sdkManifestFilePath = latestSDKManifestPath()
           if BOFFileSystemManager.isFileExist(atPath: sdkManifestFilePath ?? "") {
                var removeError: Error? = nil
               BOFFileSystemManager.removeFile(fromLocationPath: sdkManifestFilePath ?? "", removalError: removeError)
            }
            var error: Error?
           
            //else file write operation and prapare new object
           
           try BOFFileSystemManager.pathAfterWriting(contentString: sdkManifest, toFilePath: sdkManifestFilePath)//, writingError: &error)
          //  BOFFileSystemManager.path(afterWriting: sdkManifest, toFilePath: sdkManifestFilePath, writingError: &error)
            
            return sdkManifestFilePath
        } catch {
            BOFLogDebug(frmt: "%@:%@", args: BOF_DEBUG, error.localizedDescription)
        }
        return nil
    }
    
    func fetchAndPrepareSDKModel(with callback: @escaping ((_ isSuccess: Bool, _ error: Error?) -> Void)) {
        
        let api = BOManifestAPI()
        
        api.getManifestModel { responseObject in
            
            if (responseObject == nil) {
                self.isSyncedNow = false
                callback(false, nil)
                return
            }
            
            self.manifestModel = responseObject
            
            //TODO: change the storage methods used here
            
            do{
                let manifestData = try JSONEncoder().encode(responseObject)
                let manifestJSONStr = String(data: manifestData , encoding: .utf8)
                self.sdkManifestPath(afterWriting: manifestJSONStr ?? "")
                self.isSyncedNow = true
                callback(true, nil)
            }
            catch
            {
                self.isSyncedNow = false
                callback(false, error)
            }
            
            
        } failure: { error in
            self.isSyncedNow = false
            callback(false, error)
        }

    }
    
    
    func getManifestVariableModel(manifest:ManifestModel,  forID: Int)-> ManifestVariableModel?
    {
        //TODO:test this
        if let match = manifest.variables.first( where: { forID == Int($0.variableId)} ) {
            return match
        }
        return nil
    }
    
    func getManifestVariable(_ manifest: BOASDKManifest, forID ID: Int) -> BOASDKVariable? {
            var oneVar: BOASDKVariable? = nil
            if let variables = manifest.variables {
                for oneVariableDict in variables {
//                    guard let oneVariableDict = oneVariableDict else {
//                        continue
//                    }
                    if oneVariableDict != nil && oneVariableDict.variableID?.intValue == ID {
                        oneVar = oneVariableDict
                        break
                    }
                }
            }
            return oneVar
    }
    
    func reloadManifestData() {
       do{
            let manifestStr = latestSDKManifestJSONString()
            if manifestStr == nil {
                return
            }
           
           if manifestModel == nil {
               return
           }
           
           //TODO: remove this code later, looks unreachable
//           if manifestModel == nil && (manifestStr != ""){
//
//               let sdkManifestM = try BOASDKManifest.fromJSON(json: manifestStr, encoding: String.Encoding.utf8, error: manifestReadError)
//               manifestModel = sdkManifestM
//           }
           
           let systemEvents = getManifestVariableModel(manifest: manifestModel!, forID: MANIFEST_SYSTEM_EVENTS)
           if let systemEvents = systemEvents {
               enabledSystemEvents = systemEvents.value.components(separatedBy: ",")
           }

           /*
            let piiKey = getManifestVariable(sdkManifestModel!, forID: MANIFEST_PII_PUBLIC_KEY)
            if let piiKey = piiKey {
                piiPublicKey = piiKey.value
            }
            
            let phiKey = getManifestVariable(sdkManifestModel!, forID: MANIFEST_PHI_PUBLIC_KEY)
            if let phiKey = phiKey {
                phiPublickey = phiKey.value
            }
            */
            
        } catch {
            BOFLogDebug(frmt: "%@:%@", args: BOF_DEBUG, error.localizedDescription)
        }
    }
    
    func isManifestAvailable() -> Bool {

            let manifestStr = latestSDKManifestJSONString()
            if manifestStr != nil && (manifestStr != "") {
                return true
            }
            return false
    }
    
    func isSystemEventEnabled(_ eventCode: Int) -> Bool {
        if enabledSystemEvents == nil {
            return false
        }
        return ((enabledSystemEvents?.contains(NSNumber(value: eventCode).stringValue)) != nil)
    }
    
}
